### PR TITLE
Reconcile DIC only if DataSource is not managed by another DIC

### PR DESCRIPTION
**Pull request by arnongilboa**
Signed-off-by: Arnon Gilboa <agilboa@redhat.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by awels** [comment]
So what is the problem this is trying to solve. This seems like a very imperative approach. Why can't two DiCs coexist at the same time, and one of them notices the data source is owned by another DiC, and simply doesn't do anything (emit an event saying this so things observing us can see something is not right in the cluster).

**Comment by arnongilboa** [comment]
> So what is the problem this is trying to solve. This seems like a very imperative approach. Why can't two DiCs coexist at the same time, and one of them notices the data source is owned by another DiC, and simply doesn't do anything (emit an event saying this so things observing us can see something is not right in the cluster).

Why would you allow a second DIC be created and "coexist" (doing nothing) with the active one? it may have totally different source registry url, different storage etc. and add nothing but user confusion which can be eliminated on creation time. If you want them to coexist it complicates life even more as you may expect that if the active one is deleted the other one is activated.

File: pkg/controller/dataimportcron-controller.go [hunk 1]
```diff
@@ -114,12 +119,43 @@ func (r *DataImportCronReconciler) Reconcile(ctx context.Context, req reconcile.
...
+	otherCron := &cdiv1.DataImportCron{}
+	if err := r.client.Get(ctx, types.NamespacedName{Namespace: cron.Namespace, Name: dataSourceCronLabel}, otherCron); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
+	}
+	if otherCron.Spec.ManagedDataSource == dataSource.Name {
+		msg := fmt.Sprintf(MessageDataSourceAlreadyManaged, dataSource.Name, otherCron.Name)
+		r.recorder.Event(cron, corev1.EventTypeWarning, ErrDataSourceAlreadyManaged, msg)
```
Comment (awels) [inline pkg/controller/dataimportcron-controller.go:153]: besides the event, I would also log this at V(3) in the cdi controller log.

File: pkg/controller/dataimportcron-controller_test.go [hunk 2]
```diff
@@ -273,6 +273,42 @@ var _ = Describe("All DataImportCron Tests", func() {
...
+			_, err = reconciler.Reconcile(context.TODO(), cronReq)
+			Expect(err).ToNot(HaveOccurred())
+
+			dataSource = &cdiv1.DataSource{}
+			err = reconciler.client.Get(context.TODO(), dataSourceKey(cron), dataSource)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(dataSource.Labels[common.DataImportCronLabel]).To(Equal(cron.Name))
+
+			cron1 := newDataImportCron(cronName + "1")
+			shouldReconcile, err = reconciler.shouldReconcileCron(context.TODO(), cron1)
```
Comment (awels) [inline pkg/controller/dataimportcron-controller_test.go:293]: Can you check that the event is fired here? The reconciler should have an event slice you can check.

**Comment by awels** [comment]
a few minor comments but looks good otherwise.

File: pkg/controller/dataimportcron-controller.go (hunk 1 above)
Comment (arnongilboa) [reply pkg/controller/dataimportcron-controller.go:153]: added

File: pkg/controller/dataimportcron-controller_test.go (hunk 2 above)
Comment (arnongilboa) [reply pkg/controller/dataimportcron-controller_test.go:293]: added

**Pull request merged**